core/backend/src/TextAnnotationLayout.ts

```diff
@@ -349,7 +360,43 @@ export class RunLayout {
     });
   }
 
-  public toResult(paragraph: Paragraph): RunLayoutResult {
+  /**
+   * Computes offsets for each grapheme in a text run. Relies on JavaScript's `Intl.Segmenter` to detect individual graphemes.
+   * Node in the mobile add-on does not include Intl. Right now, mobile users are not expected to do any editing.
+   * Long term we will need to find a better solution.
+   * @param layoutContext
+   */
+  private computeGraphemeOffsets(layoutContext: LayoutContext): GraphemeOffset[] {
+    if (this.source.type !== "text") {
+      return [];
+    }
+
+    const content = this.stringify();
+    const baselineShift = this.source.baselineShift;
+    const style = layoutContext.createRunSettings(this.source);
+    // TypeScript only provides type declarations for Intl.Segmenter if targeting ES2022+. Cast to any until ES2022+ is supported.
+    const segmenter = new (global as any).Intl.Segmenter(undefined, {granularity: "grapheme"});
+    const graphemes: IntlSegmenterStandInType[] = Array.from(segmenter.segment(content));
+    const graphemeOffsets: GraphemeOffset[] = [];
+    let processedText = "";
+    let prevGraphemeOffset = 0;
+
+    graphemes.forEach((grapheme, index) => {
+      const nextGraphemeCharIndex = graphemes[index+1]?.index ?? content.length;
+      processedText += content.substring(grapheme.index, nextGraphemeCharIndex);
+      const rangeForText = layoutContext.computeRangeForText(processedText, style, baselineShift);
```

Comment:
Ask @jffmarker if you actually need to include all the preceding characters to get an accurate range for the current grapheme. It's a lot of redundant processing (e.g., for a 6-character run we must cumulatively process 21 characters).

Reply:
I would be concerned about glyph shaping and TrueType kerning features. For example:

![image](https://github.com/user-attachments/assets/e7a1076a-3848-4ae4-98dc-303b36bbb680)
(Word with Aptos (Body) font)

The leading edge of the second character is dependent upon the previous character (notice the 'e' is slightly tucked under the 'T', but not so for the B). I don't have good examples off-hand, but other script languages such as Devanagari and Arabic need to shape glyphs based on the preceding, so I would also expect it's important there.

![image](https://github.com/user-attachments/assets/c7649501-865c-434b-b990-7fd530cba131)
(https://github.com/n8willis/opentype-shaping-documents/blob/master/opentype-shaping-devanagari.md)

Then again, that also influences the trailing position of the previous glyph, so you may have to consider backfilling as well. Seems a shame the browser can't give a list of grapheme positions, and you need to rely on incrementally computing it like this.

Reply:
I tried it both ways (considering all prior text vs. considering only each grapheme). When considering only each grapheme, the computed offset would become incorrect if glyph shaping/kerning caused a dependency on the previous grapheme (for example, if "Te" was encountered, the cursor would appear in the middle of following graphemes, not on their edges).